server/src/main/java/io/deephaven/server/arrow/ArrowFlightUtil.java

```diff
@@ -704,7 +704,7 @@ private void apply(final BarrageSubscriptionRequest subscriptionRequest) {
             }
 
             @Override
-            public void close() {
+            synchronized public void close() {
```

Comment:
This looks accidental?

Reply:
@nbauernfeind This is supposed to fix a race condition that we run into and was making the server NPE:

The NPE happens in ArrowFlightUtil.java:674
```
                for (final BarrageSubscriptionRequest request : preExportSubscriptions) {
                    apply(request);
                }
```

The log in the server / docker compose:
```
server_1      | 2022-05-03T02:56:37.578Z | heduler-Concurrent-1 | ERROR | i.d.s.s.SessionState      | Internal Error 'e477ab1b-07a1-42c5-971b-2e1d68c1b728' io.grpc.StatusRuntimeException: CANCELLED: call already cancelled. Use ServerCallStreamObserver.setOnCancelHandler() to disable this exception
server_1      |         at io.grpc.Status.asRuntimeException(Status.java:526)
server_1      |         at io.grpc.stub.ServerCalls$ServerCallStreamObserverImpl.onNext(ServerCalls.java:366)
server_1      |         at io.deephaven.server.barrage.BarrageStreamGenerator$SnapshotView.forEachStream(BarrageStreamGenerator.java:453)
server_1      |         at io.deephaven.server.arrow.ArrowModule$1.onNext(ArrowModule.java:43)
server_1      |         at io.deephaven.server.arrow.ArrowModule$1.onNext(ArrowModule.java:38)
server_1      |         at io.deephaven.server.arrow.FlightServiceGrpcImpl.lambda$doGetCustom$6(FlightServiceGrpcImpl.java:197)
server_1      |         at io.deephaven.server.session.SessionState$ExportBuilder.lambda$submit$2(SessionState.java:1299)
server_1      |         at io.deephaven.server.session.SessionState$ExportObject.doExport(SessionState.java:847)
server_1      |         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
server_1      |         at java.util.concurrent.FutureTask.run(FutureTask.java:264)
server_1      |         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
server_1      |         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
server_1      |         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
server_1      |         at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:157)
server_1      |         at java.lang.Thread.run(Thread.java:829)
server_1      | 
server_1      | 2022-05-03T02:56:37.606Z | heduler-Concurrent-4 |  INFO | viousInstrumentedListener | BarrageMessageProducer.Operation(172923097):  Snapshot Function elapsed time 28 ms, step=9
server_1      | 2022-05-03T02:56:37.607Z | heduler-Concurrent-4 |  INFO | i.d.s.a.ArrowFlightUtil   | DoExchangeMarshaller{32631245}: processing initial subscription
server_1      | 2022-05-03T02:56:37.610Z | heduler-Concurrent-4 |  INFO | .b.BarrageMessageProducer | BarrageMessageProducer(421e49e3): Sub{3d6eea79}: subbing to columns {0-2}
server_1      | 2022-05-03T02:56:37.610Z | heduler-Concurrent-4 |  INFO | .b.BarrageMessageProducer | BarrageMessageProducer(421e49e3): Sub{3d6eea79}: scheduling update immediately, for initial snapshot.
server_1      | 2022-05-03T02:56:37.611Z | heduler-Concurrent-4 | ERROR | i.d.s.s.SessionState      | Internal Error 'cc0cd672-cf16-4ed2-b0b4-a593a707f1f4' java.lang.NullPointerException: null
server_1      |         at io.deephaven.server.arrow.ArrowFlightUtil$DoExchangeMarshaller$SubscriptionRequestHandler.onExportResolved(ArrowFlightUtil.java:674)
server_1      |         at io.deephaven.server.arrow.ArrowFlightUtil$DoExchangeMarshaller$SubscriptionRequestHandler.lambda$handleMessage$0(ArrowFlightUtil.java:624)
server_1      |         at io.deephaven.server.session.SessionState$ExportBuilder.lambda$submit$2(SessionState.java:1299)
server_1      |         at io.deephaven.server.session.SessionState$ExportObject.doExport(SessionState.java:847)
server_1      |         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
server_1      |         at java.util.concurrent.FutureTask.run(FutureTask.java:264)
server_1      |         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
server_1      |         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
server_1      |         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
server_1      |         at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:157)
server_1      |         at java.lang.Thread.run(Thread.java:829)
server_1      | 
server_1      | 2022-05-03T02:56:37.613Z | heduler-Concurrent-3 |  INFO | viousInstrumentedListener | 1109281251: Snapshot candidate step=9, rows={0-9}/{0-9}, cols={0-2}, usePrev=false
server_1      | 2022-05-03T02:56:37.617Z | heduler-Concurrent-3 | ERROR | .b.BarrageMessageProducer | Internal Error '711f748c-f38e-4957-82df-1270ce7a19d6' java.lang.IllegalStateException: Stream was terminated by error, no further calls are allowed
server_1      |         at com.google.common.base.Preconditions.checkState(Preconditions.java:510)
server_1      |         at io.grpc.stub.ServerCalls$ServerCallStreamObserverImpl.onNext(ServerCalls.java:374)
server_1      |         at io.deephaven.server.barrage.BarrageStreamGenerator$SchemaView.forEachStream(BarrageStreamGenerator.java:503)
server_1      |         at io.deephaven.server.arrow.ArrowModule$1.onNext(ArrowModule.java:43)
server_1      |         at io.deephaven.server.arrow.ArrowModule$1.onNext(ArrowModule.java:38)
server_1      |         at io.deephaven.server.barrage.BarrageMessageProducer.propagateSnapshotForSubscription(BarrageMessageProducer.java:1544)
server_1      |         at io.deephaven.server.barrage.BarrageMessageProducer.updateSubscriptionsSnapshotAndPropagate(BarrageMessageProducer.java:1417)
server_1      |         at io.deephaven.server.barrage.BarrageMessageProducer$UpdatePropagationJob.run(BarrageMessageProducer.java:992)
server_1      |         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
server_1      |         at java.util.concurrent.FutureTask.run(FutureTask.java:264)
server_1      |         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
server_1      |         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
server_1      |         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
server_1      |         at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:157)
server_1      |         at java.lang.Thread.run(Thread.java:829)
server_1      | 
```

In the client:
```
cfs@erke 22:55:42 ~/dh/oss2/deephaven-core/cpp-examples/do_exchange/build
$ ./do_exchange
Hi, ThreadNubbin is awake
/a0/h/cfs/dhcpp/src/arrow/cpp/src/arrow/flight/client.cc:363: Close() failed: Invalid: gRPC returned precondition failed error, with message: Details Logged w/ID 'cc0cd672-cf16-4ed2-b0b4-a593a707f1f4'. gRPC client debug context: {"created":"@1651546597.613389698","description":"Error received from peer ipv6:[::1]:10000","file":"/a0/h/cfs/dhcpp/src/grpc/src/core/lib/surface/call.cc","file_line":905,"grpc_message":"Details Logged w/ID 'cc0cd672-cf16-4ed2-b0b4-a593a707f1f4'","grpc_status":9}. Client context: OK
Callback reported failure: Status: Invalid: gRPC returned precondition failed error, with message: Details Logged w/ID 'cc0cd672-cf16-4ed2-b0b4-a593a707f1f4'. gRPC client debug context: {"created":"@1651546597.613389698","description":"Error received from peer ipv6:[::1]:10000","file":"/a0/h/cfs/dhcpp/src/grpc/src/core/lib/surface/call.cc","file_line":905,"grpc_message":"Details Logged w/ID 'cc0cd672-cf16-4ed2-b0b4-a593a707f1f4'","grpc_status":9}. Client context: OK. Caller message: fsr_->Next()@/a0/h/cfs/dh/oss2/deephaven-core/cpp-client/deephaven/client/src/highlevel/impl/table_handle_impl.cc:546
ThreadNubbin is exiting. Bye.
```

Reply:
LGTM

Reply:
Can we swap the keywords? i'm not aware of anywhere else in the codebase (or java at large) where `synchronized` comes before `public`

Reply:
Done.